src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -24,12 +26,344 @@
+static List * RemoveElementsUptoNode(List *waitingTransactionList,
+									 TransactionNode *transactionNode);
+static void ResetVisitedFields(HTAB *adjacencyList);
+static void AssocateDistributedTransactionWithBackendProc(TransactionNode *
+														  transactionNode);
 static TransactionNode * GetOrCreateTransactionNode(HTAB *adjacencyList,
 													DistributedTransactionId *
 													transactionId);
 static uint32 DistributedTransactionIdHash(const void *key, Size keysize);
-static int DistributedTransactionIdCompare(const void *a, const void *b, Size keysize);
+static int DistributedTransactionIdCompareHash(const void *a, const void *b, Size
+											   keysize);
+static int DistributedTransactionIdCompare(const void *a, const void *b);
+static void LogKillingBackend(TransactionNode *transactionNode);
+static void LogTransactionNode(TransactionNode *transactionNode);
+static void LogAdjacencyLists(HTAB *adjacencyLists);
+static void LogDistributedDeadlockDebugMessage(const char *errorMessage);
+
+
+PG_FUNCTION_INFO_V1(check_distributed_deadlocks);
+
+
+/*
+ * check_distributed_deadlocks is the external API for manually
+ * checking for distributed deadlocks. For the details, see
+ * CheckForDistributedDeadlocks().
+ */
+Datum
+check_distributed_deadlocks(PG_FUNCTION_ARGS)
+{
+	bool deadlockFound = CheckForDistributedDeadlocks();
+
+	return BoolGetDatum(deadlockFound);
+}
+
+
+/*
+ * CheckForDistributedDeadlocks is the entry point for detecing
+ * distributed deadlocks.
+ *
+ * In plain words, the function first builds a wait graph by
+ * adding the wait edges from the local node and then adding the
+ * remote wait edges to form a global wait graph. Later, the wait
+ * graph is converted into another graph representation (adjacency
+ * lists) for more efficient searches. Finally, a DFS is done on
+ * the adjacency lists. Finding a cycle in the graph unveils a
+ * distributed deadlock. Upon finding a deadlock, the youngest
+ * participant backend is killed.
+ *
+ * The function returns true if a deadlock is found. Otherwise, returns
+ * false.
+ */
+bool
+CheckForDistributedDeadlocks(void)
+{
+	WaitGraph *waitGraph = BuildGlobalWaitGraph();
+	HTAB *adjacencyLists = BuildAdjacencyListsForWaitGraph(waitGraph);
+	HASH_SEQ_STATUS status;
+	TransactionNode *transactionNode = NULL;
+
+	LogDistributedDeadlockDebugMessage("Distributed deadlock detection starts");
+	LogAdjacencyLists(adjacencyLists);
+
+	/*
+	 * We iterate on transaction nodes and search for deadlocks where the
+	 * starting node is the given transaction node.
+	 */
+	hash_seq_init(&status, adjacencyLists);
+	while ((transactionNode = (TransactionNode *) hash_seq_search(&status)) != 0)
+	{
+		bool deadlockFound = false;
+		List *deadlockPath = NIL;
+
+		/* we're only interested in finding deadlocks originating from this node */
+		if (transactionNode->transactionId.initiatorNodeIdentifier != GetLocalGroupId())
```

Comment:
Should store GetLocalGroupId() in a local variable, lest it's called quite a number of times.